File: kakao.app/js/ked/edit/act/multi.js
var _k_ = {list: function (l) {return l != null ? typeof l.length === 'number' ? l : [] : []}, isArr: function (o) {return Array.isArray(o)}, empty: function (l) {return l==='' || l===null || l===undefined || l!==l || typeof(l) === 'object' && Object.keys(l).length === 0}}

import kxk from "../../../kxk.js"
let kseg = kxk.kseg

import belt from "../tool/belt.js"

export default {allCursors:function ()
{
    return this.s.cursors.asMutable()
},expandCursors:function (dir)
{
    var c, cursors, dy, mc, newCursors

    cursors = this.allCursors()
    dy = (dir === 'up' ? -1 : 1)
    newCursors = []
    var list = _k_.list(cursors)
    for (var _a_ = 0; _a_ < list.length; _a_++)
    {
        c = list[_a_]
        newCursors.push(c)
        newCursors.push([c[0],c[1] + dy])
    }
    mc = belt.traversePositionsInDirection(newCursors,this.mainCursor(),dir)
    return this.setCursors(newCursors,{main:mc,adjust:'topBotDelta'})
},contractCursors:function (dir)
{
    var add, c, cursors, nbdn, nbup, newCursors, solo

    cursors = this.allCursors()
    newCursors = []
    var list = _k_.list(cursors)
    for (var _b_ = 0; _b_ < list.length; _b_++)
    {
        c = list[_b_]
        nbup = belt.positionsContain(cursors,belt.positionInDirection(c,'down'))
        nbdn = belt.positionsContain(cursors,belt.positionInDirection(c,'up'))
        solo = !(nbup || nbdn)
        add = ((function ()
        {
            switch (dir)
            {
                case 'up':
                    return nbup || solo

                case 'down':
                    return nbdn || solo

            }

        }).bind(this))()
        if (add)
        {
            newCursors.push(c)
        }
    }
    return this.setCursors(newCursors)
},addCursor:function (x, y)
{
    var cursors, pos

    pos = belt.pos(x,y)
    cursors = this.allCursors()
    cursors.push(pos)
    return this.setCursors(cursors,{main:-1})
},addCursors:function (cursors)
{
    return this.setCursors(this.allCursors().concat(cursors))
},delCursorsInRange:function (rng)
{
    var outside

    outside = belt.positionsOutsideRange(this.allCursors(),rng)
    outside.push(belt.endOfRange(rng))
    return this.setCursors(outside,{main:-1})
},moveCursors:function (dir, opt)
{
    var c, cursors, ind, line, lines, _100_22_, _99_18_

    if (_k_.isArr(dir))
    {
        switch (dir[0])
        {
            case 'bos':
                if (this.moveCursorsToStartOfSelections())
                {
                    return
                }
                dir = dir.slice(1)
                break
            case 'eos':
                if (this.moveCursorsToEndOfSelections())
                {
                    return
                }
                dir = dir.slice(1)
                break
        }

        dir = dir[0]
    }
    opt = (opt != null ? opt : {})
    opt.count = ((_99_18_=opt.count) != null ? _99_18_ : 1)
    opt.jumpWords = ((_100_22_=opt.jumpWords) != null ? _100_22_ : false)
    if (this.s.highlights.length)
    {
        this.deselect()
    }
    cursors = this.allCursors()
    lines = this.allLines()
    var list = _k_.list(cursors)
    for (var _c_ = 0; _c_ < list.length; _c_++)
    {
        c = list[_c_]
        line = lines[c[1]]
        switch (dir)
        {
            case 'left':
            case 'right':
                c[0] += belt.numCharsFromPosToWordOrPunctInDirection(lines,c,dir,opt)
                break
            case 'up':
                c[1] -= opt.count
                break
            case 'down':
                c[1] += opt.count
                break
            case 'eol':
                c[0] = kseg.width(this.s.lines[c[1]])
                break
            case 'bol':
                c[0] = 0
                break
            case 'bof':
                c[0] = 0
                c[1] = 0
                break
            case 'eof':
                c[1] = this.s.lines.length - 1
                c[0] = kseg.width(line)
                break
            case 'ind':
                c[0] = belt.numIndent(line)
                break
            case 'ind_eol':
                ind = belt.numIndent(line)
                c[0] = (c[0] < ind ? ind : kseg.width(line))
                break
            case 'ind_bol':
                ind = belt.numIndent(line)
                c[0] = (c[0] > ind ? ind : 0)
                break
        }

    }
    this.setCursors(cursors,{main:this.s.main,adjust:'topBotDelta'})
    return true
},moveCursorsToStartOfSelections:function ()
{
    var rngs, selections

    selections = this.allSelections()
    if (_k_.empty(selections))
    {
        return
    }
    rngs = belt.splitLineRanges(this.allLines(),selections,false)
    this.setCursors(belt.startPositionsOfRanges(rngs))
    return true
},moveCursorsToEndOfSelections:function ()
{
    var rngs, selections

    selections = this.allSelections()
    if (_k_.empty(selections))
    {
        return
    }
    rngs = belt.splitLineRanges(this.allLines(),selections,false)
    this.setCursors(belt.endPositionsOfRanges(rngs))
    return true
},moveCursorsToEndOfLines:function ()
{
    var cur, cursors, lines

    cursors = this.allCursors()
    lines = this.allLines()
    var list = _k_.list(cursors)
    for (var _d_ = 0; _d_ < list.length; _d_++)
    {
        cur = list[_d_]
        cur[0] = belt.lineRangeAtPos(lines,cur)[2]
    }
    this.setCursors(cursors)
    return true
},isAnyCursorInLine:function (y)
{
    var c

    var list = _k_.list(this.allCursors())
    for (var _e_ = 0; _e_ < list.length; _e_++)
    {
        c = list[_e_]
        if (c[1] === y)
        {
            return true
        }
    }
},wordAtCursor:function ()
{
    return belt.wordAtPos(this.s.lines,this.mainCursor())
},chunkBeforeCursor:function ()
{
    return belt.chunkBeforePos(this.s.lines,this.mainCursor())
},chunkAfterCursor:function ()
{
    return belt.chunkAfterPos(this.s.lines,this.mainCursor())
},setMainCursorAndSelect:function (x, y)
{
    this.setSelections(belt.extendLineRangesFromPositionToPosition,this.allLines(),this.allSelections(),this.mainCursor(),[x,y])
    return this.setCursors([[x,y]],{adjust:'topBotDelta'})
},moveCursorsAndSelect:function (dir, opt)
{
    var cursors, selections

    var _f_ = belt.extendLineRangesByMovingPositionsInDirection(this.allLines(),this.allSelections(),this.allCursors(),dir,opt); selections = _f_[0]; cursors = _f_[1]

    this.setSelections(selections)
    return this.setCursors(cursors,{adjust:'topBotDelta'})
}}

File: kakao.app/love/lua/edit/act/multi.lua
--[[
00     00  000   000  000      000000000  000    
000   000  000   000  000         000     000    
000000000  000   000  000         000     000    
000 0 000  000   000  000         000     000    
000   000   0000000   0000000     000     000    
--]]

-- use ../../../kxk ▪ kseg
-- use ../tool      ◆ belt


local multi = class("multi")
    

--  0000000   000      000      
-- 000   000  000      000      
-- 000000000  000      000      
-- 000   000  000      000      
-- 000   000  0000000  0000000  


function multi:allCursors() 
    return self.s.cursors.asMutable()
    end

-- 00000000  000   000  00000000    0000000   000   000  0000000    
-- 000        000 000   000   000  000   000  0000  000  000   000  
-- 0000000     00000    00000000   000000000  000 0 000  000   000  
-- 000        000 000   000        000   000  000  0000  000   000  
-- 00000000  000   000  000        000   000  000   000  0000000    


function multi:expandCursors(dir) 
        local cursors = self:allCursors()
        
        local dy = (function () 
    if (dir == 'up') then 
    return -1 else 
    return 1
             end
end)()
        
        local newCursors = array()
        for c in cursors do 
            newCursors:push(c)
            newCursors:push(array(c[1], (c[2] + dy)))
        end
        
        local mc = belt.traversePositionsInDirection(newCursors, self:mainCursor(), dir)
        return self:setCursors(newCursors, {main = mc, adjust = 'topBotDelta'})
    end


function multi:contractCursors(dir) 
        local cursors = self:allCursors()
        
        local newCursors = array()
        for ci, c in ipairs(cursors) do 
            local nbup = belt.positionsContain(cursors, belt.positionInDirection(c, 'down'))
            local nbdn = belt.positionsContain(cursors, belt.positionInDirection(c, 'up'))
            local solo = not (nbup or nbdn)
            local add = (function () 
    if (dir == 'up') then 
    return (nbup or solo)
                  elseif (dir == 'down') then 
    return (nbdn or solo)
                  end
end)()
            
            if add then 
                newCursors:push(c)
            end
        end
        
        return self:setCursors(newCursors)
    end

--  0000000   0000000    0000000    
-- 000   000  000   000  000   000  
-- 000000000  000   000  000   000  
-- 000   000  000   000  000   000  
-- 000   000  0000000    0000000    


function multi:addCursor(x, y) 
        local pos = belt.pos(x, y)
        local cursors = self:allCursors()
        cursors:push(pos)
        return self:setCursors(cursors, {main = -1})
    end


function multi:addCursors(cursors) 
        return self:setCursors(self:allCursors().concat(cursors))
    end


function multi:delCursorsInRange(rng) 
        local outside = belt.positionsOutsideRange(self:allCursors(), rng)
        outside:push(belt.endOfRange(rng))
        return self:setCursors(outside, {main = -1})
    end

-- 00     00   0000000   000   000  00000000  
-- 000   000  000   000  000   000  000       
-- 000000000  000   000   000 000   0000000   
-- 000 0 000  000   000     000     000       
-- 000   000   0000000       0      00000000  


function multi:moveCursors(dir, opt) 
        if is(dir, array) then 
            if (dir[0] == 'bos') then 
                    if self:moveCursorsToStartOfSelections() then return end
                    dir = dir:slice(1)
            elseif (dir[0] == 'eos') then 
                    if self:moveCursorsToEndOfSelections() then return end
                    dir = dir:slice(1)
            end
            
            dir = dir[0]
        end
        
        opt = opt or ({})
        opt.count = opt.count or 1
        opt.jumpWords = opt.jumpWords or false
        
        if (#self.s.highlights > 0) then 
            self:deselect()
        end
        
        local cursors = self:allCursors()
        local lines = self.s.lines
        
        for ci, c in ipairs(cursors) do 
            local line = lines[c[2]]
            
            if (dir == 'left') or (dir == 'right') then c[0] = c[0] + (belt.numCharsFromPosToWordOrPunctInDirection(lines, c, dir, opt))
            elseif (dir == 'up') then c[1] = c[1] - (opt.count)
            elseif (dir == 'down') then c[1] = c[1] + (opt.count)
            elseif (dir == 'eol') then c[0] = kseg.width(self.s.lines[c[2]])
            elseif (dir == 'bol') then c[0] = 0
            elseif (dir == 'bof') then c[0] = 0 ; c[2] = 0
            elseif (dir == 'eof') then c[1] = (#self.s.lines - 1) ; c[1] = kseg.width(line)
            elseif (dir == 'ind') then c[0] = belt.numIndent(line)
            elseif (dir == 'ind_eol') then local ind = belt.numIndent(line) ; c[1] = (function () 
    if (c[1] < ind) then 
    return ind else 
    return kseg.width(line)
                                                                  end
end)()
            elseif (dir == 'ind_bol') then local ind = belt.numIndent(line) ; c[1] = (function () 
    if (c[1] > ind) then 
    return ind else 
    return 1
                                                                  end
end)()
            end
        end
        
        local main = self.s.main
        local adjust = (opt.adjust or 'topBotDelta')
        
        if (dir == 'up') or (dir == 'down') or (dir == 'left') or (dir == 'right') then 
                main = belt.indexOfExtremePositionInDirection(cursors, dir, main)
                adjust = 'topBotDeltaGrow'
        end
        
        self:setCursors(cursors, {main = main, adjust = adjust})
        
        return true
    end


function multi:moveCursorsToStartOfSelections() 
        local selections = self:allSelections()
        
        if empty(selections) then return end
        
        local rngs = belt.splitLineRanges(self.s.lines, selections, false)
        
        self:setCursors(belt.startPositionsOfRanges(rngs))
        
        return true
    end


function multi:moveCursorsToEndOfSelections() 
        local selections = self:allSelections()
        
        if empty(selections) then return end
        
        local rngs = belt.splitLineRanges(self.s.lines, selections, false)
        
        self:setCursors(belt.endPositionsOfRanges(rngs))
        
        return true
    end


function multi:moveCursorsToEndOfLines() 
        local cursors = self:allCursors()
        
        for i, cur in ipairs(cursors) do 
            cur[2] = belt.lineRangeAtPos(self.s.lines, cur)[3]
        end
        
        self:setCursors(cursors)
        
        return true
    end


function multi:isAnyCursorInLine(y) 
        for i, c in ipairs(self:allCursors()) do 
            if (c[2] == y) then return true end
        end
    end

--  0000000  000   000  000   000  000   000   0000000  0000000    00000000  00000000   0000000   00000000   00000000  
-- 000       000   000  0000  000  000  000   000       000   000  000       000       000   000  000   000  000       
-- 000       000000000  000 0 000  0000000    0000000   0000000    0000000   000000    000   000  0000000    0000000   
-- 000       000   000  000  0000  000  000        000  000   000  000       000       000   000  000   000  000       
--  0000000  000   000  000   000  000   000  0000000   0000000    00000000  000        0000000   000   000  00000000  


function multi:chunksBeforeCursors() 
    return self.s.cursors:map(function (c) 
    return belt.chunkBeforePos(self.s.lines, c)
end)
    end

--  0000000  00000000  000      00000000   0000000  000000000  
-- 000       000       000      000       000          000     
-- 0000000   0000000   000      0000000   000          000     
--      000  000       000      000       000          000     
-- 0000000   00000000  0000000  00000000   0000000     000     


function multi:moveCursorsAndSelect(dir, opt) 
        local selections, cursors = belt.extendLineRangesByMovingPositionsInDirection(self.s.lines, self.s.selections, self.s.cursors, dir, opt)
        
        self:setSelections(selections)
        return self:setCursors(cursors, {adjust = 'topBotDelta'})
    end

return multi